### ISSUE TITLE
[YSQL] OOM when running COPY FROM on stdin

### ISSUE THREAD

**Issue by emhna** [yugabyte/yugabyte-db#5603]
`COPY FROM` query has 2 ways of taking input and creating a table based on it: file or stdin.
When data is read from stdin and that data is large in size (1gb+), the query will run into OOM issue.

This is because `COPY FROM` holds all rows in memory until end delimter (eg. `\.`) is processed.
An example query is `/bin/ysqlsh -c "COPY t FROM STDIN" < absFilePathToLargeFile`.
Similar to how regular files are processed, it would be ideal to process each line and insert into the table immediately rather than wait for all rows to be read.

**Comment by kmuthukk** [comment]
@emhna -- is this default behavior with vanilla Postgres also or YB specific?

**Comment by emhna** [comment]
@kmuthukk I have verified this is the behaviour only for YB relation types. 
When stdin is streamed from large file and copied over to a temporary table, the performance is stable and completes in seconds.
Eg. `/bin/ysqlsh -c "CREATE TEMP TABLE t (a int primary key, b int, c int, d int); COPY t FROM STDIN WITH (FORMAT CSV, HEADER)" < PathToLargeFile` 
However, when the table is regular type (YB relation), the query performance degrades.

**Comment by jaki** [comment]
I had a spare backup cluster with YCSB 1 million rows loaded, so I did `COPY usertable TO '/dev/shm/largefile'`.  Then, I spawned a psql connection to a postgres 12 node and ran
1. `CREATE TABLE usertable ...` (see schema in our YCSB docs on the website)
1. `COPY usertable FROM '/dev/shm/largefile'` (and the other variation using `STDIN`)
For COPY FROM, I noticed resident memory steadily increase, although hardly as much: from ~30MB to ~80MB.

When I loaded to a temporary table instead, as @emhna said, the memory usage is stable.  In fact, it is very stable, staying at 13604KB resident memory.

```
./bin/psql -c "CREATE TEMP TABLE tusertable (
    ycsb_key character varying(255) NOT NULL,
    field0 text,
    field1 text,
    field2 text,
    field3 text,
    field4 text,
    field5 text,
    field6 text,
    field7 text,
    field8 text,
    field9 text
); COPY tusertable FROM STDIN" </dev/shm/bigdata.csv
```

My guess is that vanilla postgres doesn't bother with the excessive memory usage because it is so small in scale anyway.  In our case, it's probably an easily noticeable issue since we probably consume more memory at the time of writing.  At the same time, I  think it is easy to work around using copy from file or copy from program, and this syntax probably exists for those who want to try sending the data interactively.  Therefore, I consider this low priority.

**Comment by kmuthukk** [comment]
@jaki -- agreed with considering this as low-pri for now.

**Comment by m-iancu** [comment]
Want to add some clarifications here:
1. this OOM issue is _mostly_ Yugabyte-specific, because the main culprit is Yugabyte's memory. That is because the previous fixes (066419a and 7c7c841) do not apply when copying from stdin yet (due to some internal implementation details). So the same issues we used to have there we _still_ have for this case.
2. This `stdin` is important, in part because both `pg_dump` and `ysql_dump` produce output (dump) files that use `COPY FROM stdin` internally (as it is more efficient and concise than generating inserts).

In the meantime, until this is fixed, as a *workaround* I wrote up a quick [python script](https://gist.github.com/m-iancu/01bb96b95684cac52c556531aecd31fe) that takes a dump file and writes a new one with the COPY commands inside it batched (by ending the COPY and starting a new one every _batch_size_ rows).
_Note: Another workaround is using the `--inserts` option for `pg_dump`/`ysql_dump`, but that will make data loading a lot slower because we won't be able to do any internal batching._

### Example usage.

#### Create sample data in source database
In source database:
```
create table test(k int, v text);
create table test2(a text, b int, c text);
create table test3(x text, y float, z int);

-- Using smaller values for testing.
insert into test select generate_series, generate_series from generate_series(1, 13300);
insert into test2 select generate_series, generate_series,generate_series from generate_series(1, 25500);
insert into test3 select generate_series, generate_series + 0.5,generate_series from generate_series(1, 37700);
```

#### Dump the data
```
pg_dump -U yugabyte -p 5433 -d source_db -n public > test.dmp
```

#### Fix up the dump file with the script above.
Download [batch_dump.py](https://gist.github.com/m-iancu/01bb96b95684cac52c556531aecd31fe). Might need to run `chmod a+x` to make it executable.
Then run:
```
./batch_dump.py -i test.dmp > test_batched.dmp
```

#### Import the data into target database
From `ysqlsh` connected to the target database run:
```
\i test_batched.dmp
```
Output should be:
```
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 300
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 500
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 1000
COPY 700
```
(since default batch size for the script is `1000`).

**Comment by emhna** [comment]
I have verified copy from stdin source type works against large files including:

- `copy table from stdin`

- `\copy table from 'path/to/file'`

- `pg_dump`/`ysql_dump` (generates `copy table from stdin` commands in dumped files)
